### ISSUE TITLE
test images push-build is broken

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes/kubernetes#87251]
/sig release
cc @claudiubelu @justaugustus @spiffxp 
filing this to track before I forget in this wall of meetings >.>

the push-all target appears to have broken shell currently which means it's probably not only broken in CI but locally ...

**Comment by claudiubelu** [comment]
Hi. I've commented on slack about this, so reposting here:

Looking at [1], it looks like that function was not defined. ``kubernetes/test/images/image-util.sh`` sources the file ``${KUBE_ROOT}/hack/lib/util.sh`` where that error originates from. The function ``kube::log::error`` is defined in ``${KUBE_ROOT}/hack/lib/logging.sh``. ``logging.sh`` is not sourced by either ``util.sh`` or ``image-util.sh``, but when I run ``make all-push WHAT=all-conformance`` in the ``kubernetes/test/images`` folder, I'm not getting that error.

[1] https://storage.googleapis.com/kubernetes-jenkins/logs/post-kubernetes-push-images/1217255975157764097/artifacts/build.log

**Comment by claudiubelu** [comment]
Got it. Basically, that function is called when ``kube::util::ensure-gnu-sed`` fails. This function basically checks if ``sed`` exists, and if it's GNU based or not. Unfortunately, it greps after GNU, but the ``sed`` that exists in the ``gcr.io/k8s-testimages/gcb-docker-gcloud:v20190906-745fed4`` image used by the Image Promoter, it's busybox: https://paste.ubuntu.com/p/4Cjd5HPrRv/

TLDR: So, because the image is based on busybox, ``sed`` is also busybox, and when ``image-util.sh`` checks for ``sed``, it says it doesn't "have" it, just because the ``--help`` doesn't have GNU in it's output. I think we can easily address this by treating the BusyBox case in the ``kube::util::ensure-gnu-sed`` function. Thoughts?

**Comment by BenTheElder** [comment]
maybe we should use a different image?
we check for gnu sed because of compatibility issues?

**Comment by claudiubelu** [comment]
Looking through the commit history, there are only 4 cases in which the GNU sed thing was mentioned / ensured, and the reason was pretty much so the scripts can run on Mac. From what I've read online, Macs have the FreeBSD version of sed, which behaves a bit differently.

``image-util.sh`` only uses sed to replace a few strings in some Dockerfile files, which is pretty standard, so I don't think the busybox version would have issues with that. Will check.

I am currently trying to build the images inside the aforementioned container image above, and hit another issue after the sed thing. ``image-util.sh`` builds the go binaries inside another ``golang:"${GOLANG_VERSION}"`` container:

    docker run --rm -it -v "${TARGET}:${TARGET}:Z" -v "${KUBE_ROOT}":/go/src/k8s.io/kubernetes:Z\
        golang:"${GOLANG_VERSION}" \
        /bin/bash -c "\
                cd /go/src/k8s.io/kubernetes/test/images/${SRC_DIR} && \
                CGO_ENABLED=0 ${arch_prefix} GOARCH=${ARCH} go build -a -installsuffix cgo --ldflags '-w' -o ${TARGET}/${SRC} ./$(dirname "${SRC}")"

 so the Image Promoter should also be able to build containers, but considering that docker CLI is already available inside the image, I think that'll be fine. 

Other than that, it also tries to mount the ``kubernetes`` folder inside the new container (let's call it N) (which is probably already mounted inside the builder container (let's call it A)). But I'm having issues with that, the new container N doesn't see any file in the mounted volume.

I'm looking into it, but if you have any ideas, that would be helpful.

**Comment by claudiubelu** [comment]
Update: I can currently build and push images from inside the ``gcr.io/k8s-testimages/gcb-docker-gcloud:v20190906-745fed4`` container, which is the same one used by the promoter. I've used the following command:

    docker run -t --privileged -v /var/run/docker.sock:/var/run/docker.sock \
        -v /workspace/kubernetes/:/workspace/kubernetes \
        -v /home/ubuntu/.docker:/root/.docker \
        --entrypoint bash gcr.io/k8s-testimages/gcb-docker-gcloud:v20190906-745fed4 \
        -c "cd kubernetes/test/images && \
            make all-push WHAT=all-conformance REGISTRY=claudiubelu" > build-logs

This was done using this PR: https://github.com/kubernetes/kubernetes/pull/87188

However, there are a few catches here, of which I'm not sure they're being done in the Image Promoter:
1. First of all, the container must be able to build and run other contaienrs, which is why the ``-v /home/ubuntu/.docker:/root/.docker`` was set.
2. The ``kubernetes`` repo must be mounted in ``/workspace/kubernetes``, which it probably is. But that's not enough, the ``kubernetes`` repo **must** also exist on the host in ``/workspace/kubernetes``. Basically, the mapping ``/workspace/kubernetes/:/workspace/kubernetes`` must be exactly like this. This is the solution to the previous problem I've had in the previous comment. Basically, ``image-util.sh`` builds the go binaries needed by the images through another ``golang`` container and mounts this: ``-v /workspace/kubernetes:/go/src/k8s.io/kubernetes:Z``. Now, docker will always mount the folder / volume from the original host, and not from the current container's files and folders.
3. It has to run as ``--privileged``, this is needed by the multiarch builds. This might already be set in the Image Promoter, but I'm not sure.

I'd like to hear some answers about 2 and 3.

**Comment by claudiubelu** [comment]
Addressed the last issue as well. Seems to be fine now, didn't encounter that flake anymore, and I've rebuilt all the conformance images 5-6 times already. Please review the PR, and let's try to find some answers to the points I've raised above.
